pyproject.toml

```diff
@@ -23,7 +23,7 @@ classifiers = [
     "Programming Language :: Python :: 3",
 ]
 dependencies = [
-    "jsonschema[format-nongpl,format_nongpl]>=4.3.0",
+    "jsonschema[format-nongpl,format_nongpl]>=3.2.0,<4",
```

Comment:
```suggestion
    "jsonschema[format-nongpl,format_nongpl]>=3.2.0",
```

Reply:
I should have left a comment... I was just forcing this as a temporary commit so that the unit tests used v3 and not v4. Turns out that doing so means two of the validation tests fail. Not sure if that means this pr can't work?

Reply:
If those two tests emit a different error on v3, I'd say update the tests to detect that error (perhaps checking the version of `jsonschema`.

Reply:
So in this commit I find that on jsconschema 3.2.0 an exception is no longer thrown where it is currently expected: https://github.com/bretttully/jupyter_events/commit/192c83f1a331dc9b38af5bf44901471ecd60e445

I am happy to disable this test for jsonschema <4, but does that make sense, or is this indicating that you need >=4?

Reply:
Hi @bretttully - I'm encountering the same issue with the `kfp` cap of `jsonschema` in my server extension so I've been following this issue.

I think the remaining issue is that the way the `jsonschema` dependency is configured, the optional dependencies in `jsonschema` relative to `format_nongpl` do not get installed and, as a result, the expected `ValidationError` instances are not thrown.

If I explicitly issue `pip install "jsonschema[format_nongpl]==3.2.0"` then the tests pass using your branch.  But using `pip install "jsonschema[format-nongpl,format_nongpl]==3.2.0"` results the following warning message:
```
WARNING: jsonschema 3.2.0 does not provide the extra 'format-nongpl'
```
and, via the results, skips the installation of other optional dependencies.  Only the first format leads to the installation of `rfc3339-validator-0.1.4` and `rfc3986-validator-0.1.1` - which are responsible for `date-format` and `uri` validation, respectively.

As a result, I believe this is now an exercise in how best to express the `jsonschema` dependency (which was [adjusted from `format_nongpl` to `format-nongpl` in its 4.9.0 release](https://github.com/python-jsonschema/jsonschema/commit/438c8fb12546d761c7ce3cba7fe36057c11fe713)).

Unfortunately, this is an area of limited knowledge for me, but I'm sure others will know how to address this.

Thank you for raising, and working on, this issue!

Reply:
Hmm - I just tried this `pip install --force-reinstall "jsonschema[format_nongpl,format-nongpl]==3.2.0"` and it did the right thing; warned that `format-nongpl` was not recognized.  Similarly, installing `==4.17.0` warned that `format_nongpl` was not recognized but still installed the optional packages.

This seems to imply that perhaps the package installation of `jupyter_events` is what can't deal with the unrecognized optional dependency.

I'm hoping @bollwyvl or Steve @blink1073 can shed some light here.

Reply:
I think `jsonschema` 3.2 is getting bit by the normalization of extras: https://discuss.python.org/t/what-extras-names-are-treated-as-equal-and-why/7614/4.

Reply:
Yeah, those warnings don't stop installation.  I think the right thing to do is skip the tests with jsonschema 3 and just accept the fact that there is less validation when using jsonschema 3.

Reply:
okay.  But I've confirmed the tests work fine provided the appropriate validation packages are present.  Probably splitting hairs but if folks require `jsonschema < 4` they won't be privy to the same validation.

Reply:
We could duplicate just those two deps I suppose.

Reply:
Can we add a package list to the CI tests?  I think you'll find `rfc3339-validator` and `rfc3986-validator` are not present when `jsonschema < 4` is installed by `jupyter_events`.

Here's the result of pip installing the wheel:
```
$ pip install dist/jupyter_events-0.6.0.dev0-py3-none-any.whl 
Processing ./dist/jupyter_events-0.6.0.dev0-py3-none-any.whl
Requirement already satisfied: pyyaml>=6.0 in /opt/miniconda3/envs/events-dev/lib/python3.9/site-packages (from jupyter-events==0.6.0.dev0) (6.0)
Collecting jsonschema[format-nongpl]<4,>=3.2.0
  Using cached jsonschema-3.2.0-py2.py3-none-any.whl (56 kB)
Requirement already satisfied: python-json-logger>=2.0.4 in /opt/miniconda3/envs/events-dev/lib/python3.9/site-packages (from jupyter-events==0.6.0.dev0) (2.0.4)
Requirement already satisfied: traitlets>=5.3 in /opt/miniconda3/envs/events-dev/lib/python3.9/site-packages (from jupyter-events==0.6.0.dev0) (5.8.0)
WARNING: jsonschema 3.2.0 does not provide the extra 'format-nongpl'
Requirement already satisfied: setuptools in /opt/miniconda3/envs/events-dev/lib/python3.9/site-packages (from jsonschema[format-nongpl]<4,>=3.2.0->jupyter-events==0.6.0.dev0) (65.6.3)
Requirement already satisfied: six>=1.11.0 in /opt/miniconda3/envs/events-dev/lib/python3.9/site-packages (from jsonschema[format-nongpl]<4,>=3.2.0->jupyter-events==0.6.0.dev0) (1.16.0)
Requirement already satisfied: pyrsistent>=0.14.0 in /opt/miniconda3/envs/events-dev/lib/python3.9/site-packages (from jsonschema[format-nongpl]<4,>=3.2.0->jupyter-events==0.6.0.dev0) (0.19.3)
Requirement already satisfied: attrs>=17.4.0 in /opt/miniconda3/envs/events-dev/lib/python3.9/site-packages (from jsonschema[format-nongpl]<4,>=3.2.0->jupyter-events==0.6.0.dev0) (22.2.0)
Installing collected packages: jsonschema, jupyter-events
Successfully installed jsonschema-3.2.0 jupyter-events-0.6.0.dev0
```

I suspect the normalization is applied by the hatch/pyproject.toml machinery, since that's the only way I can explain why the explicit `pip install jsonschema[format_nongpl]==3.2.0` works.

Reply:
> We could duplicate just those two deps I suppose.

Oh, you mean go ahead and add the format-nongpl dependencies explicitly?
(Just lost power so will be delayed in responding further)

Reply:
> add the format-nongpl dependencies explicitly?

Just the two you identified, to ensure a subset of validation is working.